draft-ietf-quic-retry-offload.md

```diff
@@ -174,43 +174,60 @@ the servers.
-Initial Packet. This has a number of considerations.
-
-* If a Retry Offload maintains no per-flow state, it cannot distinguish between
-valid and invalid non-Initial packets and MUST forward all non-Initial Packets
-to the server.
-
-* For QUIC versions the Retry Offload does not support and are present on the
-Allow-List (or absent from the Deny-List), the Retry Offload cannot distinguish
-Initial Packets from other long headers and therefore MUST admit all long
-headers.
-
-* If a Retry Offload keeps per-flow state, it can identify 4-tuples that have
-been previously approved, admit non-Initial packets from those flows, and
-drop all others. However, dropping short headers will effectively break Address
-Migration and NAT Rebinding when in Active Mode, as post-migration packets will
-arrive with a previously unknown 4-tuple. This policy will also break connection
-attempts using any new QUIC versions that begin connections with a short header.
-
-* If a Retry Offload is integrated with a QUIC-LB routable load balancer
-{{?I-D.ietf-quic-load-balancers}}, it can verify that the Destination Connection
-ID is routable, and only admit non-Initial packets with routable DCIDs. As the
-Connection ID encoding is invariant across QUIC versions, the Retry Offload can
-do this for all short headers.
-
-Nothing in this section prevents Retry Offloads from making basic syntax
-correctness checks on packets with QUIC versions that it understands (e.g.,
-enforcing the Initial Packet datagram size minimum in version 1).
+## Retry Offloads with Per-Connection State
+
+A Retry Offload that keeps per-connection state can keep track of 5-tuples that
+it has accepted, either because it was in inactive mode or because it contained
+a valid token. It SHOULD accept subsequent Initial packets from these 4-tuples,
+regardless of the presence of a token, to avoid dropping part of the client's
+second flight.
+
+With per-connection state, the Retry Offload MAY drop Handshake and 0-RTT
+packets that do not correspond to an accepted 5-tuple. It SHOULD NOT drop short
+header packets, as these may be the result of a connection migration.
+
+## Retry Offloads without Per-Connection State
+
+Without per-connection state, Retry Offloads MUST admit all datagrams that begin
+with non-Initial packets.
+
+If a client Initial packet arrives without a Retry token, it might be the second
+client packet flight of a connection that was admitted when the Retry Offload
+was in inactive mode. If the Initial Packet does not consume the entire length
+of the UDP datagram, the Retry Offload MUST identify other packet types in the
+datagram. If there is a Handshake packet present, the Retry Offload MUST remove
+the Initial Packet and MAY remove any 0-RTT packet from the datagram, updating
+the UDP payload length accordingly. This prevents connection deadlock when
+client handshake packets progress the connection and the server is subject to
+the QUIC amplification limit.
```

Comment:
I don't care for this for two reasons:

1. It requires Retry Offloads to parse the entire UDP datagram (or at least more of it). This could be leveraged to attack the Retry Offload itself.
2. It requires the Retry Offloads to modify the contents of the datagram in these cases. This text stats the UDP payload length should be updated accordingly, but it doesn't go into any details about what you actually put at the end of the datagram (after the valid packets). And again, this requirement can be leveraged to attack the Retry Offload.

Reply:
1. The limited amount of parsing seems quite inexpensive to me.

2. You're not putting anything at the end of the datagram, you're just deleting Initial and 0-RTT. Invalid hanshake packets can be immediately discarded at the server, so there are no costs to letting them through. Meanwhile, if this is the second flight the server will do fine if Initial doesn't get through.